apps/emqx_durable_storage/src/emqx_ds_bitmask_keymapper.erl

```diff
@@ -0,0 +1,746 @@
+%%   |   -<-/     |     -<-/     |     -<-/
+%%   | -/         |   -/         |   -/
+%%   | ---->------|   ---->------|   ---------->
+%%   |
+%%  -+------------+-----------------------------> t
+%%        epoch
+%%
+%% This structure allows to quickly seek to a the first message that
+%% was recorded in a certain epoch in a certain topic or a
+%% group of topics matching filter like `foo/bar/#`.
+%%
+%% Due to its structure, for each pair of rocksdb keys K1 and K2, such
+%% that K1 > K2 and topic(K1) = topic(K2), timestamp(K1) >
+%% timestamp(K2).
+%% That is, replay doesn't reorder messages published in each
+%% individual topic.
+%%
+%% This property doesn't hold between different topics, but it's not deemed
+%% a problem right now.
+%%
+%%================================================================================
+
+%% API:
+-export([
+    make_keymapper/1,
+    vector_to_key/2,
+    bin_vector_to_key/2,
+    key_to_vector/2,
+    bin_key_to_vector/2,
+    key_to_bitstring/2,
+    bitstring_to_key/2,
+    make_filter/2,
+    ratchet/2,
+    bin_increment/2,
+    bin_checkmask/2,
+    bitsize/1
+]).
+
+-export_type([vector/0, key/0, dimension/0, offset/0, bitsize/0, bitsource/0, keymapper/0]).
+
+-compile(
+    {inline, [
+        ones/1,
+        extract/2,
+        extract_inv/2
+    ]}
+).
+
+-elvis([{elvis_style, no_if_expression, disable}]).
+
+-ifdef(TEST).
+-include_lib("proper/include/proper.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-type scalar() :: integer().
+
+-type vector() :: [scalar()].
+
+%% N-th coordinate of a vector:
+-type dimension() :: pos_integer().
+
+-type offset() :: non_neg_integer().
+
+-type bitsize() :: pos_integer().
+
+%% The resulting 1D key:
+-type key() :: non_neg_integer().
+
+-type bitsource() ::
+    %% Consume `_Size` bits from timestamp starting at `_Offset`th
+    %% bit from N-th element of the input vector:
+    {dimension(), offset(), bitsize()}.
+
+-record(scan_action, {
+    src_bitmask :: integer(),
+    src_offset :: offset(),
+    dst_offset :: offset()
+}).
+
+-type scan_action() :: #scan_action{}.
+
+-type scanner() :: [[scan_action()]].
+
+-record(keymapper, {
+    schema :: [bitsource()],
+    scanner :: scanner(),
+    size :: non_neg_integer(),
+    dim_sizeof :: [non_neg_integer()]
+}).
+
+-opaque keymapper() :: #keymapper{}.
+
+-type scalar_range() :: any | {'=', scalar() | infinity} | {'>=', scalar()}.
+
+-include("emqx_ds_bitmask.hrl").
+
+-type filter() :: #filter{}.
+
+%%================================================================================
+%% API functions
+%%================================================================================
+
+%% @doc Create a keymapper object that stores the "schema" of the
+%% transformation from a list of bitsources.
+%%
```

Comment:
Maybe it's worth mentioning that the order of bitsources is relevant, so that the first bitsource will be placed adjacent to the rightmost (LSB?) bits, the next one to the "left" of the previous result, etc.?

Reply:
Done.